src/nexmark/queries/q14.rs

```diff
@@ -0,0 +1,139 @@
+use super::NexmarkStream;
+use crate::{nexmark::model::Event, operator::FilterMap, Circuit, OrdZSet, Stream};
+use rust_decimal::Decimal;
+
+/// Query 14: Calculation (Not in original suite)
+///
+/// Convert bid timestamp into types and find bids with specific price.
+/// Illustrates duplicate expressions and usage of user-defined-functions.
+///
+/// ```sql
+/// CREATE FUNCTION count_char AS 'com.github.nexmark.flink.udf.CountChar';
+///
+/// CREATE TABLE discard_sink (
+///     auction BIGINT,
+///     bidder BIGINT,
+///     price  DECIMAL(23, 3),
+///     bidTimeType VARCHAR,
+///     dateTime TIMESTAMP(3),
+///     extra VARCHAR,
+///     c_counts BIGINT
+/// ) WITH (
+///   'connector' = 'blackhole'
+/// );
+///
+/// INSERT INTO discard_sink
+/// SELECT
+///     auction,
+///     bidder,
+///     0.908 * price as price,
+///     CASE
+///         WHEN HOUR(dateTime) >= 8 AND HOUR(dateTime) <= 18 THEN 'dayTime'
+///         WHEN HOUR(dateTime) <= 6 OR HOUR(dateTime) >= 20 THEN 'nightTime'
+///         ELSE 'otherTime'
+///     END AS bidTimeType,
+///     dateTime,
+///     extra,
+///     count_char(extra, 'c') AS c_counts
+/// FROM bid
+/// WHERE 0.908 * price > 1000000 AND 0.908 * price < 50000000;
+/// ```
+
+#[derive(Eq, Clone, Debug, PartialEq, PartialOrd, Ord)]
+enum BidTimeType {
+    Day,
+    Night,
+    Other,
+}
+
+#[derive(Eq, Clone, Debug, PartialEq, PartialOrd, Ord)]
+pub struct Q14Output(u64, u64, Decimal, BidTimeType, u64, String, usize);
+
+type Q14Stream = Stream<Circuit<()>, OrdZSet<Q14Output, isize>>;
+
+// This is used because we can't currently use chrono.Utc, which would simply
+// be Utc.timestamp_millis(b.date_time as i64).hour(), as it's waiting on a
+// release to fix a security issue.
+fn hour_for_millis(millis: u64) -> usize {
+    let millis_for_day = millis % (24 * 60 * 60 * 1000);
+    (millis_for_day / (60 * 60 * 1000)) as usize
+}
+
+pub fn q14(input: NexmarkStream) -> Q14Stream {
+    input.flat_map(|event| match event {
+        Event::Bid(b) => {
+            let new_price = Decimal::new((b.price * 100) as i64, 2) * Decimal::new(908, 3);
+            match usize::try_from(new_price).unwrap() {
+                1_000_000..=50_000_000 => Some(Q14Output(
```

Comment:
The SQL query has strict inequalities (`<`, `>`)

Reply:
Yeah, it does, but our `price` field on the model itself is an integer (as in the original), so I can't represent or test that boundary condition (1_000_000.0 / 0.908 = 1_101_321.6). But right, that shouldn't stop me from representing it correctly.